texar/modules/pretrained/xlnet_utils.py

```diff
@@ -0,0 +1,247 @@
+                           for part in "qkv"]
+            assign(layer.head_projection.weight,
+                   torch.cat([
+                       p.view(p.size(0), -1) for p in qkv_weights
+                   ], dim=1).t())
+            assign_linear(layer.pos_projection, prefix + "r/")
+            assign(layer.output_projection.weight,  # DO NOT TRANSPOSE THIS!!!!
+                   prefix + "o/kernel", lambda p: p.view(p.size(0), -1))
+            assign_layer_norm(layer.layer_norm, prefix)
+
+        for idx in range(n_layers):
+            layer: PositionWiseFF = xlnet.ff_layers[idx]
+            prefix = f"transformer/layer_{idx}/ff/"
+            for linear_idx in range(1, 2 + 1):
+                linear_prefix = f"{prefix}layer_{linear_idx}/"
+                linear_layer: nn.Linear = getattr(layer, f"linear{linear_idx}")
+                assign_linear(linear_layer, linear_prefix)
+            assign_layer_norm(layer.layer_norm, prefix)
+
+        seg_embeds = [
+            p.squeeze(0)
+            for p in torch.chunk(
+                get_weight("transformer/seg_embed"), n_layers, dim=0)]
+        for idx in range(n_layers):
+            assign(xlnet.attn_layers[idx].segment_embed, seg_embeds[idx])
+
+        if hasattr(xlnet, 'mask_emb') and hasattr(xlnet, 'lm_bias'):
+            assign(xlnet.mask_emb, "transformer/mask_emb/mask_emb")
+            assign(xlnet.lm_bias, "lm_loss/bias")
+
+    load_xlnet_model(model)
+
+    if len(from_params) > 0:
+        print(f"WARNING: Certain weights from checkpoint are not loaded: "
+              f"{list(from_params.keys())}")
+
+    filtered_to_params = [k for k in to_params if k.startswith("xlnet")]
+    if len(filtered_to_params) > 0:
+        print(f"WARNING: Certain parameters are not initialized: "
+              f"{list(filtered_to_params)}")
+
+
+def _default_download_dir() -> str:
+    r"""Return the directory to which packages will be downloaded by default.
+    """
+    package_dir = os.path.dirname(os.path.dirname(
+        os.path.dirname(os.path.dirname(__file__))))
+    if os.access(package_dir, os.W_OK):
+        texar_download_dir = os.path.join(package_dir, 'texar_download')
+    else:
+        # On Windows, use %APPDATA%
+        if sys.platform == 'win32' and 'APPDATA' in os.environ:
+            home_dir = os.environ['APPDATA']
+
+        # Otherwise, install in the user's home directory.
+        else:
+            home_dir = os.path.expanduser('~/')
+            if home_dir == '~/':
+                raise ValueError("Could not find a default download directory")
+
+        texar_download_dir = os.path.join(home_dir, 'texar_download')
+
+    if not os.path.exists(texar_download_dir):
+        os.mkdir(texar_download_dir)
+
+    return os.path.join(texar_download_dir, 'xlnet')
+
+
+def load_pretrained_model(pretrained_model_name: str,
+                          cache_dir: Optional[str] = None) -> str:
```

Comment:
And this method also. Let's pass the filename and/or cache path as arguments.

Reply:
I prefer to keep these functions (`load_pretrained_bert`, `load_pretrained_gpt2`, and `load_pretrained_xlnet`) since there are some slight differences (mainly depending on how the model authors store the checkpoint files) among these functions.